Tools/Scripts/libraries/webkitflaskpy/webkitflaskpy/ipc_manager.py

```diff
@@ -0,0 +1,122 @@
+# Copyright (C) 2023 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import hashlib
+import json
+
+from webkitcorepy import string_utils
+from webkitflaskpy import Database
+
+
+class IPCManager(object):
+     DEFAULT_EXPIRATION = 10 * 60
+     DEFAULT_TIMEOUT = 60
+     RETRY = 3
+
+     def __init__(self, *databases):
+         self.databases = databases or [Database()]
+         self.callbacks = dict()
+
+     def register(self, name, callback):
+         if not isinstance(name, str):
+             raise ValueError('Name of registered callback must be a string')
+         if not callable(callback):
+             raise ValueError('Callback must be callable')
+         self.callbacks[name] = callback
+
+     def enqueue(self, name, *args, **kwargs):
+         expiration = kwargs.pop('ex', self.DEFAULT_EXPIRATION)
+         retry = kwargs.pop('retry', self.RETRY)
+         if name not in self.callbacks:
+             raise ValueError("'{}' is not a registered callback".format(name))
+         encoded = json.dumps([name, args, kwargs, retry, expiration])
+         digest = hashlib.md5()
+         digest.update(string_utils.encode(encoded))
+         digest = digest.hexdigest()
+
+         for database in self.databases:
+             database.set(digest, encoded, ex=expiration)
```

Comment:
Just to confirm, `database.set` guarantees the atomic operation here?

Reply:
Inside Redis, yes. The function call isn't atomic (we do some mutating of the digest so give it a partition key), but that actual insertion into the database is.